File: src/main/java/com/javarush/island/shevchenko/services/WorldLifeCycle.java
package com.javarush.island.shevchenko.services;

import com.javarush.island.shevchenko.entities.islandMap.Location;
import com.javarush.island.shevchenko.settings.IslandSettings;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;


/**
 * Класс WorldLifeCycle представляет собой жизненный цикл симуляции на острове животных.
 * Он отвечает за запуск, остановку и управление шагами симуляции.
 */


public class WorldLifeCycle {

    private final WorldInitializer world;
    private final AtomicInteger simulationStepNumber = new AtomicInteger(0);
    private final ScheduledExecutorService service = Executors.newSingleThreadScheduledExecutor();

    public WorldLifeCycle(WorldInitializer worldGenerator) {
        this.world = worldGenerator;
    }

    private void startingTheSimulation() {
        Location[][] locations = world.getIslandField().getLocations();
        simulationStepNumber.incrementAndGet();
        world.getIslandField().output();
        for (int y = 0; y < locations[y].length; y++) {
            for (Location[] location : locations) {
                location[y].start();
            }
        }
    }


    private final Runnable lifeTask = new Runnable() {
        @Override
        public void run() {
            Location[][] locations = world.getIslandField().getLocations();
            startingTheSimulation();
            int numberOfSimulationSteps = IslandSettings.NUMBER_OF_SIMULATION_STEPS;
            if (simulationStepNumber.get() >= numberOfSimulationSteps) {
                service.shutdown();
                for (int y = 0; y < locations[y].length; y++) {
                    for (Location[] location : locations) {
                        location[y].shutdown();
                    }
                }
                for (int y = 0; y < locations[y].length; y++) {
                    for (Location[] location : locations) {
                        try {
                            location[y].await(500);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                }
                world.getIslandField().output();
            }
        }
    };


    public void start() {
        int stepDuration = IslandSettings.STEP_DURATION;
        service.scheduleAtFixedRate(lifeTask, 1, stepDuration, TimeUnit.MILLISECONDS);
    }
}

File: src/main/kotlin/ru/cooper/island/services/WorldLifeCycle.kt
package ru.cooper.island.services

import ru.cooper.island.settings.IslandSettings
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

/**
 * Класс WorldLifeCycle представляет собой жизненный цикл симуляции на острове животных.
 * Он отвечает за запуск, остановку и управление шагами симуляции.
 */
class WorldLifeCycle(private val world: WorldInitializer) {
    private val simulationStepNumber = AtomicInteger(0)
    private val service: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()

    private fun startingTheSimulation() {
        val locations = world.islandField.locations
        simulationStepNumber.incrementAndGet()
        world.islandField.output()
        for (y in locations[y].indices) {
            for (location in locations) {
                location[y]!!.start()
            }
        }
    }

    private val lifeTask = Runnable {
        val locations = world.islandField.locations
        startingTheSimulation()
        val numberOfSimulationSteps = IslandSettings.NUMBER_OF_SIMULATION_STEPS
        if (simulationStepNumber.get() >= numberOfSimulationSteps) {
            service.shutdown()
            for (y in locations[y].indices) {
                for (location in locations) {
                    location[y]!!.shutdown()
                }
            }
            for (y in locations[y].indices) {
                for (location in locations) {
                    try {
                        location[y]!!.await(500)
                    } catch (e: InterruptedException) {
                        e.printStackTrace()
                    }
                }
            }
            world.islandField.output()
        }
    }

    fun start() {
        val stepDuration = IslandSettings.STEP_DURATION
        service.scheduleAtFixedRate(lifeTask, 1, stepDuration.toLong(), TimeUnit.MILLISECONDS)
    }
}
